doc/developer/design/20230329_monotonic_optimization.md

```diff
@@ -0,0 +1,177 @@
+- Feature name: Monotonic rendering of SELECT queries
+- Associated: (Insert list of associated epics, issues, or PRs)
+
+# Summary
+[summary]: #summary
+
+<!--
+One paragraph to explain the feature: What's wrong, what's the problem, what's the fix, what are the consequences?
+-->
+
+At the moment, we render dataflows for `SELECT` queries in the same way as dataflows serving indexes and materialized views.  This prevents us from using monotonic operators, which by design have a lower runtime memory footprint.  Instead, we want to exploit the fact that a `SELECT` queries the data at a single timestamp, which means that all data can be interpreted as monotonic.  With this change, we can render `SELECT` queries with more monotonic operators and we can teach the optimizer to distinguish several monotonicity variants.
+
+# Motivation
+[motivation]: #motivation
+
+<!--
+Why are we doing this? What problems does it solve? What use cases does it enable? What is the desired outcome?
+-->
+
+Materialize currently focuses on rendering a good plan for workloads that need incremental view maintenance (IVM).  This misses opportunities to optimize plans based on the time bounds of a dataflow, which are in the case for `SELECT` queries a single timestamp.  In this design, we present an updated approach for the optimizer to handle monotonicity information, and to tighten the contract between plans and how they're rendered.
+
+# Explanation
+[explanation]: #explanation
+
+<!--
+Explain the design as if it were part of Materialize and you were teaching the team about it.
+This can mean:
+
+- Introduce new named concepts.
+- Explain the feature using examples that demonstrate product-level changes.
+- Explain how it builds on the current architecture.
+- Explain how engineers and users should think about this change, and how it influences how everyone uses the product.
+- If needed, talk though errors, backwards-compatibility, or migration strategies.
+- Discuss how this affects maintainability, or whether it introduces concepts that might be hard to change in the future.
+-->
+
+To start, we introduce the following monotonicity variants:
+* *Non-monotonic*: The data can contain retractions over time.
+* *Physically monotonic*: The data never contains retractions.
+* *Logically monotonic*: The data can contain retractions, but once consolidated it is physically monotonic.
+
+At the moment, the optimizer thinks in terms of the first two variants, but rendering assumes logical monotonicity.
```

Comment:
I think rendering relies on physical monotonicity as there are some operators that are not correct without it. Some operator implementations would be correct with logical monotonicity, and have the opportunity to have simpler implementations with physical monotonicity.

Reply:
I think a good first step when tackling this is to extend this design doc (or some related sub-issue) with some kind of table that reflects which operators assume / require which type of monotonicity at the moment, and how we would like to evolve them in the future.